### DISCUSSION TITLE
SetFilterMTMode (Avisynth wiki)

### DISCUSSION THREAD

**Discussion by Asd-g** [AviSynth/AviSynthPlus#382] [General]
IMO [the Avisynth wiki page of SetFilterMTMode](http://avisynth.nl/index.php/SetFilterMTMode) should have explanation of the behavior of `SetFilterMTMode(.., bool "force")`. Something like: SetFilterMTMode(.., bool "force") can change the MT modes of the filters in the chain before the filter that has SetFilterMTMode(.., bool "force"). And some examples:

```
SetFilterMTMode(X, MT_SERIALIZED, force=true)
sourceFilter

# filterY has own MT mode MT_MULTI_INSTANCE but now is changed to MT_SERIALIZED
filterY

# filterZ has own MT mode MT_NICE_FILTER but now is changed to MT_SERIALIZED
filterZ

filterX
```

Another example:

```
SetFilterMTMode(X, MT_MULTI_INSTANCE, force=true)
sourceFilter

# filterY has own MT mode MT_SERIALIZED, not changed
filterY

# filterZ has own MT mode MT_NICE_FILTER but now is changed to MT_MULTI_INSTANCE
filterZ

filterX
```

Another example:

```
SetFilterMTMode(X, MT_NICE_FILTER, force=true)
sourceFilter

# filterY has own MT mode MT_SERIALIZED, not changed
filterY

# filterZ has own MT mode MT_MULTI_INSTANCE, not changed
filterZ

filterX
```

Another example:

```
SetFilterMTMode(X, MT_SERIALIZED, force=true)
SetFilterMTMode(Y, MT_NICE_FILTER, force=true)
sourceFilter

# filterY has own MT mode MT_MULTI_INSTANCE but now is changed to MT_SERIALIZED
# SetFilterMTMode(Y, MT_NICE_FILTER, force=true) doesn't have effect
filterY

# filterZ has own MT mode MT_NICE_FILTER but now is changed to MT_SERIALIZED
filterZ

filterX
```

Another example:

```
SetFilterMTMode(X, MT_MULTI_INSTANCE, force=true)
SetFilterMTMode(Y, MT_SERIALIZED, force=true)
sourceFilter

# filterY has own MT mode MT_MULTI_INSTANCE but now is changed to MT_SERIALIZED
filterY

# filterZ has own MT mode MT_NICE_FILTER but now is changed to MT_MULTI_INSTANCE
filterZ

filterX
```

Also some links can be added - [this post](https://forum.doom9.org/showthread.php?p=1823754#post1823754) that is reply of [this one](https://forum.doom9.org/showthread.php?p=1823726#post1823726).

Also some info for `SetFilterMTMode(..., force=false)` - this has effect only if the filter doesn't set own MT mode. Usually the recent versions of the plugins have set MT mode.

Mainly tagging @pinterf.

**Comment by pinterf** [comment]
Thanks, sooner or later all documentation will be maintained in our rst documentation. 
I've just updated this MT part yesterday, but I can insert these examples and links.

But there is no resource to maintain avisynth.nl wiki version parallel, but I can imagine some backports later.

As for MT_SERIALIZED setting can alter the MT behaviour of other filters in the chain: 
http://avisynth.nl/index.php/SetFilterMTMode

I included some more changes and your examples.

Current version (wait until it gets rebuilt after the commits)
https://avisynthplus.readthedocs.io/en/latest/avisynthdoc/syntax/syntax_internal_functions_multithreading_new.html

**Comment by Asd-g** [comment]
Thanks for updating the rst documentation. I will backport the info to avisynth.nl wiki.

I noted `<todo check: what's the effect of using multiple Prefetch sections?>` in the rst documentation. Yes, as a workaround `prefetch` can be used for every filter in the chain before the one with `MT_SERIALIZED`. For example:

```
sourceFilter

# filterY has own MT mode MT_MULTI_INSTANCE
filterY
prefetch(2)

# filterZ has own MT mode MT_NICE_FILTER
filterZ
prefetch(2)

# filterX has own MT mode MT_SERIALIZED
filterX

**Reply by pinterf** [reply]
I put there a hint, mentioning MtModeEvaluator, partially for myself, to check it again. Perhaps we should stop tracing back MtModeEvaluator until Prefetch is found?

**Reply by Asd-g** [reply]
Sorry, no idea (I cannot figure out the workflow).

Edit: I tried to add some info to the avisynth.nl wiki but it seems I don't rights to edit http://avisynth.nl/index.php/SetFilterMTMode

**Comment by pinterf** [comment]
I've successfully edited the content; inserted the usual link to the updated documentation. I was asked about what AVI meant and that's all.

**Reply by Asd-g** [reply]
I clicked on the "Save" button but nothing happened. Maybe my browser (uBlock) blocks that captcha.
Thanks for the added info.
